### ISSUE TITLE
Using of JasksonFeature in combination with Jersey and Grizzly in SSL mode may lead to corrupted data in HTTPS request

### ISSUE THREAD

**Issue by netudima** [eclipse-ee4j/glassfish-grizzly#2007]
Symptoms:  when we send a large enough POJO as a JSON REST request body via Jersey with Grizzly Connector with enabled SSL - we receive a corrupted request body on a server side.

The problem is reproduced when all of the following conditions are met:
- HTTPS  is used
- Jackson body writer is used to convert POJO to HTTP request body (org.glassfish.jersey.jackson.JacksonFeature enabled)
- Request POJO is large enough (serialized size > 8096 bytes)

Libraries versions:
- Jersey: 2.27
- Grizzly: 2.4.3
- Jackson: 2.8.10

Older versions are affected as well.

Test to reproduce: [corruptedRequestTest.zip](https://github.com/eclipse-ee4j/grizzly/files/2020564/corruptedRequestTest.zip)

Root cause analysis:
The following sequence of circumstances triggers the problem:
1) An applied code invokes Jersey REST client to send an REST request with a POJO as a body.
2) As a part of the request sending Jersey uses Jackson writer extension (com.fasterxml.jackson.databind.ObjectWriter) to convert the POJO to a JSON byte stream to write HTTP request body.
3) Internally ObjectWriter uses UTF8JsonGenerator (com.fasterxml.jackson.core.json.UTF8JsonGenerator) which one is actually doing the serialization to JSON.
4) UTF8JsonGenerator uses an internal byte array (_outputBuffer) as buffer to accumulate a portion of serialized data and flushes it periodically to an underline entity output stream. The buffer typical size is 8000.
```java
protected final void _flushBuffer() throws IOException
{
    int len = _outputTail;
    if (len > 0) {
        _outputTail = 0;
        _outputStream.write(_outputBuffer, 0, len);
    }
}
```
Note, the _outputBuffer is re-used by UTF8JsonGenerator after writing to the stream.

5) Output stream which Jersey provides to write entity delegates to com.ning.http.client.providers.grizzly.FeedableBodyGenerator.Feeder. This delegation is established in org.glassfish.jersey.grizzly.connector.GrizzlyConnector.translate() logic (by setting a StreamProvider).
Feeder is used to split the input stream to HTTP chunks and trigger the underline HTTP client (Grizzly) to transfer chunks. The passed byte arrays are wrapped (not copied) to ByteBuffer instances.
6) Grizzly HTTP client is asynchronous one and may queue request messages in some cases. It looks like (contract is not defined clearly, so it is questionable) messages are expected to be cloned in this case, if it is needed.
Some evidence from org.glassfish.grizzly.asyncqueue.MessageCloner class:
```java
**
 * Cloner, which will be called by {@link AsyncQueueWriter}, when message
 * could not be written directly, and will be added to the queue.
 * Cloner may create a clone of original message and return it to the
 * {@link AsyncQueueWriter} instead of original one.
 * Using MessageCloner, developer has a chance to clone a message only in
 * case, when it is really required.
*/
```
There is also a similar discussion here: http://grizzly.1045725.n5.nabble.com/Grizzly-http-client-race-condition-td5710847.html 

Because SSL is enabled we have SSLFilter (org.glassfish.grizzly.ssl.SSLFilter) as a part of filter chain. SSLFilter may accumulates original messages in an internal org.glassfish.grizzly.ssl.SSLFilter.SSLHandshakeContext#pendingWriteContexts list. It happens when we want to send a request and HTTP client needs to establish a new SSL/TLS session, in this case SSLFilter executes SSL Handshake asynchronously and while it is in progress - SSLFilter accumulates messages to send and passes them to a IO writer when SSL handshake is completed.

In summary we have a case when a re-usable byte array is sent by UTF8JsonGenerator to a REST library output stream. The array is passed via several layers to Grizzly network library where SSLFilter in some cases may put it into a queue and process it later when another portion of data is passed to the network library. The buffer is not copied during the processing, just wrapped into some containers like ByteBuffer. So, when UTF8JsonGenerator will write new data into his buffer after a flush - it may corrupt some previously flushed data if they are not send out yet (SSL handshake is in progress).

**Comment by netudima** [comment]
Work arounds for this problem:

Option 1: Enable BUFFERED instead of default CHUNKED option:
```java
clientBuilder.property(ClientProperties.REQUEST_ENTITY_PROCESSING, RequestEntityProcessing.BUFFERED);
```
The option activates another logic branch in org.glassfish.jersey.grizzly.connector.GrizzlyConnector.translate() which fully buffers the entity stream to a non-reusable byte array before sending it to Grizzly.
This option disables HTTP Chunks and may be not tolerant to very large requests (it allocates an byte array with a full request body).

Option 2: Copy byte array which is flushed to an entity output stream before passing it to Grizzly HTTP client. It can be implemented using of javax.ws.rs.ext.WriterInterceptor extension point in Jersey to replace entity stream with a copying wrapper:
```java
clientBuilder.register(new CopyingStreamWriterInterceptor())

public static class CopyingStreamWriterInterceptor implements WriterInterceptor {
    @Override
    public void aroundWriteTo(WriterInterceptorContext context) throws IOException, WebApplicationException {
        OutputStream outputStream = context.getOutputStream();
        CopyingOutputStream copyingOutputStream = new CopyingOutputStream(outputStream);
        context.setOutputStream(copyingOutputStream);
        context.proceed();
    }
}

public static class CopyingOutputStream extends OutputStream {
    private final OutputStream delegate;

    public CopyingOutputStream(OutputStream delegate) {
        this.delegate = delegate;
    }

    @Override
    public void write(int b) throws IOException {
        delegate.write(b);
    }

    @Override
    public void write(byte[] b) throws IOException {
        byte[] copy = Arrays.copyOf(b, b.length);
        delegate.write(copy);
}

    @Override
    public void write(byte[] b, int off, int len) throws IOException {
        byte[] copy = Arrays.copyOfRange(b, off, len);
        delegate.write(copy);
    }

    @Override
    public void close() throws IOException {
        delegate.close();
    }
}
```
